codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -925,5 +915,295 @@ final void incrementStreamableBytesForTree(int numBytes) {
+
+        /**
+         * Attempt to allocate bytes to streams which have frames queued.
+         * @throws Http2Exception If a write occurs and an exception happens in the write operation.
+         */
+        public abstract void writePendingBytes() throws Http2Exception;
+
+        /**
+         * Called when the writability of the underlying channel changes.
+         * @throws Http2Exception If a write occurs and an exception happens in the write operation.
+         */
+        public void channelWritabilityChange() throws Http2Exception { }
+
+        /**
+         * Called when the state is cancelled outside of a write operation.
+         * @param state the state that was cancelled.
+         */
+        public void stateCancelled(AbstractState state) { }
+
+        /**
+         * Increment the window size for a particular stream.
+         * @param state the state associated with the stream whose window is being incremented.
+         * @param delta The amount to increment by.
+         * @throws Http2Exception If this operation overflows the window for {@code state}.
+         */
+        public void incrementWindowSize(AbstractState state, int delta) throws Http2Exception {
+            state.incrementStreamWindow(delta);
+        }
+
+        /**
+         * Add a frame to be sent via flow control.
+         * @param state The state associated with the stream which the {@code frame} is associated with.
+         * @param frame the frame to enqueue.
+         * @throws Http2Exception If a writability error occurs.
+         */
+        public void enqueueFrame(AbstractState state, FlowControlled frame) throws Http2Exception {
+            state.enqueueFrame(frame);
+        }
+
+        /**
+         * Increment the total amount of pending bytes for all streams. When any stream's pending bytes changes
+         * method should be called.
+         * @param delta The amount to increment by.
+         */
+        public final void incrementPendingBytes(int delta) {
+            totalPendingBytes += delta;
+
+            // Notification of writibilty change should be delayed until the end of the top level event.
+            // This is to ensure the flow controller is more consistent state before calling external listener methods.
+        }
+
+        /**
+         * Determine if the stream associated with {@code state} is writable.
+         * @param state The state which is associated with the stream to test writability for.
+         * @return {@code true} if {@link AbstractState#stream()} is writable. {@code false} otherwise.
+         */
+        public final boolean isWritable(AbstractState state) {
+            return isWritableConnection() && state.windowSize() - state.pendingBytes() > 0;
+        }
+
+        protected final void writePendingBytes(Http2StreamVisitor writeAllocatedBytesVisitor) throws Http2Exception {
+            if (ctx == null) {
```

Comment:
@nmittler - Yes good point. I guess for the remote side this should not be necessary. I was thinking the user may want to control this before the channel is setup/connected.